homeassistant/components/jellyfin/media_source.py

```diff
@@ -0,0 +1,308 @@
+    MEDIA_CLASS_ALBUM,
+    MEDIA_CLASS_ARTIST,
+    MEDIA_CLASS_DIRECTORY,
+    MEDIA_CLASS_TRACK,
+)
+from homeassistant.components.media_player.errors import BrowseError
+from homeassistant.components.media_source.models import (
+    BrowseMediaSource,
+    MediaSource,
+    MediaSourceItem,
+    PlayMedia,
+)
+from homeassistant.core import HomeAssistant
+
+from .const import (  # pylint:disable=unused-import
+    COLLECTION_TYPE_MUSIC,
+    DATA_CLIENT,
+    DOMAIN,
+    ITEM_KEY_COLLECTION_TYPE,
+    ITEM_KEY_ID,
+    ITEM_KEY_IMAGE_TAGS,
+    ITEM_KEY_INDEX_NUMBER,
+    ITEM_KEY_MEDIA_SOURCES,
+    ITEM_KEY_MEDIA_TYPE,
+    ITEM_KEY_NAME,
+    ITEM_TYPE_ALBUM,
+    ITEM_TYPE_ARTIST,
+    ITEM_TYPE_AUDIO,
+    ITEM_TYPE_LIBRARY,
+    MAX_STREAMING_BITRATE,
+    MEDIA_SOURCE_KEY_PATH,
+    MEDIA_TYPE_AUDIO,
+    MEDIA_TYPE_NONE,
+    SUPPORTED_COLLECTION_TYPES,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_get_media_source(hass: HomeAssistant) -> MediaSource:
+    """Set up Jellyfin media source."""
+    # Currently only a single Jellyfin server is supported
+    entry = hass.config_entries.async_entries(DOMAIN)[0]
+
+    data = hass.data[DOMAIN][entry.entry_id]
+    client = data[DATA_CLIENT]
+
+    return JellyfinSource(hass, client)
+
+
+class JellyfinSource(MediaSource):
+    """Represents a Jellyfin server."""
+
+    name: str = "Jellyfin"
+
+    def __init__(self, hass: HomeAssistant, client: JellyfinClient):
+        """Initialize the Jellyfin media source."""
+        super().__init__(DOMAIN)
+
+        self.hass = hass
+
+        self.client = client
+        self.api = client.jellyfin
+        self.url = jellyfin_url(client, "")
+
+    async def async_resolve_media(self, item: MediaSourceItem) -> PlayMedia:
+        """Return a streamable URL and associated mime type."""
+        media_item = await self.hass.async_add_executor_job(
+            self.api.get_item, item.identifier
+        )
+
+        stream_url = self._get_stream_url(media_item)
+        mime_type = _media_mime_type(media_item)
+
+        return PlayMedia(stream_url, mime_type)
+
+    async def async_browse_media(
+        self, item: MediaSourceItem, media_types: Tuple[str] = None
+    ) -> BrowseMediaSource:
+        """Return a browsable Jellyfin media source."""
+        if not item.identifier:
+            return await self._build_libraries()
+
+        media_item = await self.hass.async_add_executor_job(
+            self.api.get_item, item.identifier
+        )
+
+        item_type = media_item["Type"]
+        if item_type == ITEM_TYPE_LIBRARY:
+            return await self._build_library(media_item, True)
+        if item_type == ITEM_TYPE_ARTIST:
+            return await self._build_artist(media_item, True)
+        if item_type == ITEM_TYPE_ALBUM:
+            return await self._build_album(media_item, True)
+
+        raise BrowseError("Unsupported item type %s" % item_type)
```

Comment:
https://developers.home-assistant.io/docs/development_guidelines/#use-new-style-string-formatting
```suggestion
        raise BrowseError(f"Unsupported item type {item_type}")
```